File: app/src/main/java/li/klass/fhem/appwidget/view/widget/base/DeviceAppWidgetView.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.appwidget.view.widget.base;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.os.SystemClock;
import android.util.Log;
import android.widget.RemoteViews;

import com.google.common.base.Optional;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;

import li.klass.fhem.R;
import li.klass.fhem.activities.AndFHEMMainActivity;
import li.klass.fhem.appwidget.WidgetConfiguration;
import li.klass.fhem.appwidget.WidgetConfigurationCreatedCallback;
import li.klass.fhem.appwidget.annotation.SupportsWidget;
import li.klass.fhem.appwidget.view.WidgetType;
import li.klass.fhem.constants.Actions;
import li.klass.fhem.constants.BundleExtraKeys;
import li.klass.fhem.constants.ResultCodes;
import li.klass.fhem.dagger.ForApplication;
import li.klass.fhem.domain.core.Device;
import li.klass.fhem.fragments.FragmentType;
import li.klass.fhem.service.room.RoomListService;
import li.klass.fhem.util.FhemResultReceiver;

public abstract class DeviceAppWidgetView extends AppWidgetView {

    public static final String TAG = DeviceAppWidgetView.class.getName();

    @Inject
    RoomListService roomListService;

    @Inject
    @ForApplication
    Context applicationContext;

    public static final Logger LOG = LoggerFactory.getLogger(DeviceAppWidgetView.class);

    public boolean supports(Device<?> device) {
        if (!device.getClass().isAnnotationPresent(SupportsWidget.class)) return false;

        if (!device.supportsWidget(this.getClass())) {
            return false;
        }

        SupportsWidget annotation = device.getClass().getAnnotation(SupportsWidget.class);
        Class<? extends DeviceAppWidgetView>[] supportedWidgetViews = annotation.value();
        for (Class<? extends DeviceAppWidgetView> supportedWidgetView : supportedWidgetViews) {
            if (supportedWidgetView.equals(this.getClass())) return true;
        }

        return false;
    }

    @Override
    public RemoteViews createView(Context context, WidgetConfiguration widgetConfiguration) {
        RemoteViews views = super.createView(context, widgetConfiguration);
        Log.i(TAG, "creating appwidget view for " + widgetConfiguration);

        if (shouldSetDeviceName()) {
            String deviceName = deviceNameFrom(widgetConfiguration);

            Device device = getDeviceFor(deviceName);
            if (device == null) return null;

            views.setTextViewText(R.id.deviceName, device.getWidgetName());
        }

        return views;
    }

    private String deviceNameFrom(WidgetConfiguration widgetConfiguration) {
        return widgetConfiguration.payload.get(0);
    }

    public boolean shouldSetDeviceName() {
        return true;
    }

    private Device getDeviceFor(String deviceName) {
        return roomListService.getDeviceForName(deviceName).orNull();
    }

    protected void openDeviceDetailPageWhenClicking(int viewId, RemoteViews view, Device device, WidgetConfiguration widgetConfiguration) {
        openDeviceDetailPageWhenClicking(viewId, view, device, widgetConfiguration.widgetId);
    }

    protected void openDeviceDetailPageWhenClicking(int viewId, RemoteViews view, Device device, int widgetId) {
        PendingIntent pendingIntent = createOpenDeviceDetailPagePendingIntent(device, widgetId);

        view.setOnClickPendingIntent(viewId, pendingIntent);
    }

    protected PendingIntent createOpenDeviceDetailPagePendingIntent(Device device, int widgetId) {
        Intent openIntent = createOpenDeviceDetailPageIntent(device, applicationContext);
        return PendingIntent.getActivity(applicationContext, widgetId, openIntent,
                PendingIntent.FLAG_UPDATE_CURRENT);
    }

    protected Intent createOpenDeviceDetailPageIntent(Device device, Context context) {
        Intent openIntent = new Intent(context, AndFHEMMainActivity.class);
        openIntent.setFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        openIntent.putExtra(BundleExtraKeys.FRAGMENT, FragmentType.DEVICE_DETAIL);
        openIntent.putExtra(BundleExtraKeys.DEVICE_NAME, device.getName());
        openIntent.putExtra("unique", "foobar://" + SystemClock.elapsedRealtime());
        return openIntent;
    }

    @Override
    public void createWidgetConfiguration(Context context, WidgetType widgetType, int appWidgetId,
                                          WidgetConfigurationCreatedCallback callback, String... payload) {
        Optional<Device> device = roomListService.getDeviceForName(payload[0]);
        if (device.isPresent()) {
            createDeviceWidgetConfiguration(context, widgetType, appWidgetId, device.get(), callback);
        } else {
            Log.i(TAG, "cannot find device for " + payload[0]);
        }
    }

    protected void createDeviceWidgetConfiguration(Context context, WidgetType widgetType, int appWidgetId,
                                                   Device device, WidgetConfigurationCreatedCallback callback) {
        callback.widgetConfigurationCreated(new WidgetConfiguration(appWidgetId, widgetType, device.getName()));
    }

    protected void fillWidgetView(Context context, RemoteViews view,
                                  WidgetConfiguration widgetConfiguration) {
        Device<?> device = getDeviceFor(deviceNameFrom(widgetConfiguration));
        if (device != null) {
            view.setTextViewText(R.id.deviceName, device.getWidgetName());
            fillWidgetView(context, view, device, widgetConfiguration);
        } else {
            Log.i(TAG, "cannot find device for " + deviceNameFrom(widgetConfiguration));
        }
    }

    protected abstract void fillWidgetView(Context context, RemoteViews view, Device<?> device,
                                           WidgetConfiguration widgetConfiguration);
}

File: app/src/main/java/li/klass/fhem/appwidget/ui/widget/base/DeviceAppWidgetView.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.appwidget.ui.widget.base

import android.app.PendingIntent
import android.content.Context
import android.widget.RemoteViews
import androidx.navigation.NavDeepLinkBuilder
import li.klass.fhem.R
import li.klass.fhem.activities.AndFHEMMainActivity
import li.klass.fhem.adapter.devices.core.detail.DeviceDetailRedirectFragmentArgs
import li.klass.fhem.appwidget.ui.widget.WidgetConfigurationCreatedCallback
import li.klass.fhem.appwidget.update.WidgetConfiguration
import li.klass.fhem.domain.core.FhemDevice
import li.klass.fhem.update.backend.DeviceListService
import li.klass.fhem.update.backend.device.configuration.DeviceConfigurationProvider
import li.klass.fhem.util.toHtml
import org.slf4j.LoggerFactory
import javax.inject.Inject

abstract class DeviceAppWidgetView : AppWidgetView() {

    @Inject
    lateinit var deviceListService: DeviceListService

    @Inject
    lateinit var deviceConfigurationProvider: DeviceConfigurationProvider

    open fun supports(device: FhemDevice): Boolean =
            supportsFromJsonConfiguration(device)

    private fun supportsFromJsonConfiguration(device: FhemDevice): Boolean {
        val configuration = deviceConfigurationProvider.configurationFor(device)
        val supportedWidgets = configuration.supportedWidgets
        return supportedWidgets
                .any { javaClass.simpleName.equals(it, ignoreCase = true) }
    }

    override fun createView(context: Context, widgetConfiguration: WidgetConfiguration): RemoteViews {
        val views = super.createView(context, widgetConfiguration)
        logger.info("createView - creating appwidget view for $widgetConfiguration")

        if (shouldSetDeviceName()) {
            val deviceName = deviceNameFrom(widgetConfiguration)

            val device = getDeviceFor(deviceName, widgetConfiguration.connectionId)
            if (device == null) {
                logger.info("createView - device is null, ignoring update", device)
            } else {
                views.setTextViewText(R.id.deviceName, device.widgetName.toHtml())
            }
        }

        return views
    }

    fun deviceNameFrom(widgetConfiguration: WidgetConfiguration): String =
            widgetConfiguration.payload[0]

    open fun shouldSetDeviceName(): Boolean = true

    private fun getDeviceFor(deviceName: String, connectionId: String?): FhemDevice? =
            deviceListService.getDeviceForName(deviceName, connectionId)

    protected fun openDeviceDetailPageWhenClicking(viewId: Int, view: RemoteViews, device: FhemDevice, widgetConfiguration: WidgetConfiguration, context: Context) {
        val pendingIntent = createOpenDeviceDetailPageIntent(device, widgetConfiguration, context)

        view.setOnClickPendingIntent(viewId, pendingIntent)
    }

    fun createOpenDeviceDetailPageIntent(device: FhemDevice, widgetConfiguration: WidgetConfiguration, context: Context): PendingIntent =
            NavDeepLinkBuilder(context)
                    .setComponentName(AndFHEMMainActivity::class.java)
                    .setGraph(R.navigation.nav_graph)
                    .setDestination(R.id.deviceDetailRedirectFragment)
                    .setArguments(DeviceDetailRedirectFragmentArgs(device.name, widgetConfiguration.connectionId).toBundle())
                    .createPendingIntent()

    override fun createWidgetConfiguration(context: Context, appWidgetId: Int, connectionId: String,
                                           callback: WidgetConfigurationCreatedCallback, vararg payload: String) {
        val device = deviceListService.getDeviceForName(payload[0])
        if (device != null) {
            createDeviceWidgetConfiguration(context, appWidgetId, connectionId, device, callback)
        } else {
            logger.info("cannot find device for " + payload[0])
        }
    }

    protected fun valueForMarker(device: FhemDevice, annotationCls: Class<out Annotation>): String? {
        val configuration = deviceConfigurationProvider.configurationFor(device)
        val states = configuration.states
        return states
                .filter { it.markers.contains(annotationCls.simpleName) }
                .mapNotNull { device.xmlListDevice.stateValueFor(it.key) }
                .firstOrNull()
    }

    protected open fun createDeviceWidgetConfiguration(context: Context, appWidgetId: Int, connectionId: String, device: FhemDevice,
                                                       callback: WidgetConfigurationCreatedCallback) {
        callback.widgetConfigurationCreated(WidgetConfiguration(appWidgetId, widgetType, connectionId, listOf(device.name)))
    }

    override fun fillWidgetView(context: Context, view: RemoteViews,
                                widgetConfiguration: WidgetConfiguration) {
        val device = getDeviceFor(deviceNameFrom(widgetConfiguration), widgetConfiguration.connectionId)
        if (device != null) {
            view.setTextViewText(R.id.deviceName, device.widgetName.toHtml())
            fillWidgetView(context, view, device, widgetConfiguration)
        } else {
            logger.info("cannot find device for " + deviceNameFrom(widgetConfiguration))
        }
    }

    protected abstract fun fillWidgetView(context: Context, view: RemoteViews, device: FhemDevice,
                                          widgetConfiguration: WidgetConfiguration)

    companion object {
        val logger = LoggerFactory.getLogger(DeviceAppWidgetView::class.java)!!
    }
}
